kernel/cpu_local/src/lib.rs

```diff
@@ -219,6 +219,13 @@ pub fn init<P>(
     size_of_per_cpu_data: usize,
     per_cpu_data_initializer: impl FnOnce(usize) -> P
 ) -> Result<(), &'static str> {
+    // Enable FSGSBASE instructions
+    #[cfg(target_arch = "x86_64")]
+    {
+        use x86_64::registers::control::{Cr4, Cr4Flags};
+        unsafe { Cr4::update(|flags| flags.insert(Cr4Flags::FSGSBASE)) };
+    }
+    
```

Comment:
even after reading the intel SDM, i cannot tell if this CR4 bit is only required to enable these instructions for non-privileged execution modes (ring 3) or if it's always required.
Also, is this set of instructions supported on every processor generation or just newer ones? If it's not supported on all x86_64/amd64 hardware, i'd prefer to revert to using the old MSR-based `GsBase::write()`, especially since it doesn't need to be efficient here, as we only write it once upon initializing each CPU.

There's definitely an argument to be made for using this in `set_next_task()` to update `FSBASE`, since that runs on every context switch.

Reply:
Without setting the CR4 bit, the instruction was triggering an illegal instruction exception.

> Also, is this set of instructions supported on every processor generation or just newer ones?

Well we're using Broadwell and that's pretty old. FWIW `sel4` [uses `fsgsbase` instructions by default](https://github.com/seL4/seL4/blob/49f4ddfc8c20249b7a114a4cc3baa42f5f9eaa6c/src/arch/x86/config.cmake#L210) when compiling for `x86_64`.

Reply:
Ok yea that's what I was wondering, whether it'd fault if we didn't set that bit. All good.

Reply:
> We call `rdgsbase` every time we access a CPU local with a user-defined type

True, but should we be doing that? I know that reading an MSR is slow, but perhaps using `rdgsbase` is faster than using a CLS self pointer and accessing it via `[%gs:0]`?

Reply:
Ignore this, we can discuss it on the other comment.